LocoNet monitor Ops CV test

sidlo64 opened the pull request:
@devel-bobm please help.
I don't have the hardware to verify this. Please check line 6782.

sidlo64 committed:
LocoNet monitor Ops CV test

devel-bobm commented on java/test/jmri/jmrix/loconet/messageinterp/LocoNetMessageInterpretTest.java:
```diff
@@ -6777,6 +6777,14 @@ public void testExpendedAccyReads() {
+        Assert.assertEquals("read 7 Ext Accy Addr 500 CV 260", ""
+                + "Extended Accessory Decoder CV Verify: Address 500 CV 260, check if it is 222.\n",
+                LocoNetMessageInterpret.interpretMessage(l, "LT", "LS", "LR"));
     }
```
It would be better if the code would attempt to check a few different CV numbers, _especially_ CV numbers that exercise the CV numbers that were changed.  At a minimum, the tests should exercise CV Numbers 256, 257, 512, 513, and 1024, to verify bits 8 and 9, and the maximum CV number value.

devel-bobm commented on java/test/jmri/jmrix/loconet/messageinterp/LocoNetMessageInterpretTest.java:
```diff
@@ -6809,6 +6817,20 @@ public void testExpendedAccyWrites() {
+        Assert.assertEquals("write 5: Ext Accy Addr 500 CV 260 to 222", ""
+                + "Extended Accessory Decoder CV Write: Address 500 CV 260 with 222.\n",
+                LocoNetMessageInterpret.interpretMessage(l, "LT", "LS", "LR"));
+        
```
It would be better if the code would attempt to check a few different CV numbers, _especially_ CV numbers that exercise the CV numbers that were changed.  At a minimum, the tests should exercise CV Numbers 256, 257, 512, 513, and 1024, to verify bits 8 and 9, and the maximum CV number  value.

devel-bobm commented on java/test/jmri/jmrix/loconet/messageinterp/LocoNetMessageInterpretTest.java:
```diff
@@ -6809,6 +6817,20 @@ public void testExpendedAccyWrites() {
+        Assert.assertEquals("write 6: Ext Accy Addr 500 CV 260 to 222", ""
+                + "Extended Accessory Decoder CV Write: Address 500 CV 260 with 222.\n",
+                LocoNetMessageInterpret.interpretMessage(l, "LT", "LS", "LR"));
     }
```
Note that comment that says "DT602 Extended Accessory Decoder CV" access is _specifically_ a "DT602 Extended Accessory Decoder CV access to a Digitrax 7th-generation Accessory decoder".  I would suggest that you change the comment(s) to show something like that.

sidlo64 committed:
I made tests for the Extended accessory decoder packet.

For DT602 I suggest the name Basic accessory decoder packet. It is according to NMRA 9.2.1, byte 2, bit 7 = 1.

I don't have the hardware (DT602) to Write Basic packet and Verify packet. If anyone sends me LocoNet packets from DT602, I will add them to the tests.
If not, I derive them from known data. It will be laborious, but it will be possible.

sidlo64 committed:
fix typo in test write 9

devel-bobm commented:
@sidlo64 , the link you provide does not give me specifif information of what you are concerned of, because github and my browser does not tell me what you are concerned with.  I figure you are referring to line 284 and 285 of the Run HeadlessTest with Maven.

That problem _is specifically_ because the Java code is printing with a "friendly" way of showing numbers, not a "raw" way.  The numbers between 1000 and 999999 _display_ with a "," to make the number "more readable".  Since you have a value of 1024 for the address, _and_ 1024 for the CV number, both _need_ to expect the "comma"s.

I generally try to test the "address" values _separately_ from the CV numbers, and those _separately_ from the data values, so that a mistake in one part of the code is easier to identify, especially when I mistakenly have used the wrong byte in the wrong part of the code.  

So I will have tests for bunch of Addresses, with fixed CV numbers and fixed "data" values.  Then I will have tests for a bunch of CV numbers with a fixed address and fixed "data" valuesh .  Then I will test a bunch of "data" values with a fixed address and a fixed CV number.  Yes, this requires more coding in the test, and more test time, but it does a better job of testing for the types of things that I and other coders tend to have problems with!

devel-bobm commented:
And, for completeness, "friendly" numbers (in "modern American English") typically have "comma"s every 3 places to the "left" of the "decimal point".  Like "999", "1,000", "1,500.01", "999,999.3", "1,000,000", "1,000,000,000", ...

numbers greater than or equal to one million get "2" commas, like "1,000,000".  And numbers greater than or a thousamd million (a "Billion"

sidlo64 committed:
fix address and CV 1,024

sidlo64 commented:
@devel-bobm thank you for help.
I fixed numbers 1,024.
Text changed from DT602 to Basic.

I need help with LocoNet packets from DT602 for write and read. Could you help me?

devel-bobm commented:
Here are some CV read requests, from DT602, to a Digitrax 7th-generation Accessory decoder at address 512.  I chose various CV numbers, as seen by JMRI 5.7.9, with _CV numbers_ that targeted certain "powers of two" as shown in the LocoNet data, _not_ for decimal values shown on the JMRI textual display in the log.

[ED 0B 7F 54 07 3F 6E 64 00 00 00]  Extended Accessory Decoder CV Verify: Address 512 CV 1, check if it is 0.
[ED 0B 7F 54 07 3F 6E 64 01 00 01]  Extended Accessory Decoder CV Verify: Address 512 CV 2, check if it is 0.
[ED 0B 7F 54 07 3F 6E 64 02 02 00]  Extended Accessory Decoder CV Verify: Address 512 CV 3, check if it is 2.
[ED 0B 7F 54 07 3F 6E 64 04 00 04]  Extended Accessory Decoder CV Verify: Address 512 CV 5, check if it is 0.
[ED 0B 7F 54 07 3F 6E 64 08 00 08]  Extended Accessory Decoder CV Verify: Address 512 CV 9, check if it is 0.
[ED 0B 7F 54 17 3F 6E 64 10 03 03]  Extended Accessory Decoder CV Verify: Address 512 CV 17, check if it is 131.
[ED 0B 7F 54 07 3F 6E 64 20 00 20]  Extended Accessory Decoder CV Verify: Address 512 CV 33, check if it is 0.
[ED 0B 7F 54 07 3F 6E 64 40 00 40]  Extended Accessory Decoder CV Verify: Address 512 CV 65, check if it is 0.
[ED 0B 7F 54 0F 3F 6E 64 00 00 08]  Extended Accessory Decoder CV Verify: Address 512 CV 129, check if it is 0.
[ED 0B 7F 54 07 3F 6E 65 00 00 01]  Extended Accessory Decoder CV Verify: Address 512 CV 257, check if it is 0.
[ED 0B 7F 54 07 3F 6E 66 00 00 02]  Extended Accessory Decoder CV Verify: Address 512 CV 513, check if it is 0.
[ED 0B 7F 54 0F 3F 6E 67 7F 00 74]  Extended Accessory Decoder CV Verify: Address 512 CV 1,024, check if it is 0.



Here are some CV writerequests, from DT602, to a Digitrax 7th-generation Accessory decoder at address 512.  I chose various CV numbers, as seen by JMRI 5.7.9, with _CV numbers_ that targeted certain "powers of two" as shown in the LocoNet data, _not_ for decimal values shown on the JMRI textual display in the log.


[ED 0B 7F 54 0F 3F 6E 6F 7F 00 7C]  Extended Accessory Decoder CV Write: Address 512 CV 1,024 with 0.
[ED 0B 7F 54 07 3F 6E 6E 00 01 0B]  Extended Accessory Decoder CV Write: Address 512 CV 513 with 1.
[ED 0B 7F 54 07 3F 6E 6D 00 02 0B]  Extended Accessory Decoder CV Write: Address 512 CV 257 with 2.
[ED 0B 7F 54 0F 3F 6E 6C 00 04 04]  Extended Accessory Decoder CV Write: Address 512 CV 129 with 4.
[ED 0B 7F 54 07 3F 6E 6C 40 08 40]  Extended Accessory Decoder CV Write: Address 512 CV 65 with 8.
[ED 0B 7F 54 07 3F 6E 6C 20 10 38]  Extended Accessory Decoder CV Write: Address 512 CV 33 with 16.
[ED 0B 7F 54 07 3F 6E 6C 10 20 38]  Extended Accessory Decoder CV Write: Address 512 CV 17 with 32.
[ED 0B 7F 54 07 3F 6E 6C 08 40 40]  Extended Accessory Decoder CV Write: Address 512 CV 9 with 64.
[ED 0B 7F 54 17 3F 6E 6C 04 00 1C]  Extended Accessory Decoder CV Write: Address 512 CV 5 with 128.
[ED 0B 7F 54 17 3F 6E 6C 02 7F 65]  Extended Accessory Decoder CV Write: Address 512 CV 3 with 255.
[ED 0B 7F 54 07 3F 6E 6C 01 7F 76]  Extended Accessory Decoder CV Write: Address 512 CV 2 with 127.
[ED 0B 7F 54 17 3F 6E 6C 00 7E 66]  Extended Accessory Decoder CV Write: Address 512 CV 1 with 254.

Hope this helps.  If yoy need more, just tell me what you want, specifically!

sidlo64 committed:
add LocoNet messages Digitrax 7th-generation Acc dec

sidlo64 commented:
@devel-bobm thank you for your help. I used some LocoNet messages.
Headless CI Tests passed without error. I think we are done.
